### PR TITLE
Restructure serializers. Add `serializer_headers()` and `as_attachment()`

### PR DESCRIPTION
**Pull request by schloerke**
Fixes: https://github.com/rstudio/plumber/pull/455
Fixes: #482

* Build each serializer on top of one another to avoid copy/pasta
* Add serializer_headers(headers)
* Have all text-like content return `charset=UTF-8`
* Use an `.onLoad()` method for adding the serializers. (Allows for globals to not be directly altered while the package is loading.)
* Remove all serializer files and merge into serializer.R
* Added `force` argument to allow serializers to be overwritten

PR task list:
- [x] Update NEWS
- [x] Add tests
- [x] Update documentation with `devtools::document()`
- [x] Add `@ycphs` as author to NEWS

### CODE REVIEW

File: R/serializer.R [hunk 1]
```diff
@@ -17,30 +8,267 @@ NULL
...
-addSerializer <- function(name, serializer){
-  if (is.null(.globals$serializers)){
-    .globals$serializers <- list()
-  }
-  if (!is.null(.globals$serializers[[name]])){
-    stop ("Already have a serializer by the name of ", name)
+addSerializer <- function(name, serializer, verbose = TRUE) {
+  if (!is.null(.globals$serializers[[name]])) {
+    if (isTRUE(verbose)) {
+      stop ("Already have a serializer by the name of ", name)
```
Comment (schloerke) [inline R/serializer.R:17]:
```suggestion
      warning("Overwriting serializer: ", name)
```

File: R/serializer.R [hunk 2]
```diff
@@ -17,30 +8,267 @@ NULL
...
+#' Plumber Serializers
+#'
+#' Serializers are used in Plumber to transform the R object produced by a
+#' filter/endpoint into an HTTP response that can be returned to the client. See
+#' [here](https://book.rplumber.io/articles/rendering-output.html#serializers-1) for
+#' more details on Plumber serializers and how to customize their behavior.
+#' @describeIn serializers Add a static list of headers to each return value. Will add `Content-Disposition` header if a value is the result of `as_attachment()`.
+#' @param ... extra arguments supplied to respective internal serialization function.
+#' @param headers `list()` of headers to add to the response object
+#' @param serialize_fn Function to serialize the data. The result object will be converted to a character string. Ex: `jsonlite::toJSON`.
```
Comment (cpsievert) [inline R/serializer.R:80]: Maybe try `[jsonlite::toJSON()]` with recent version of roxygen2

File: R/serializer.R [hunk 3]
```diff
@@ -17,30 +8,267 @@ NULL
...
-addSerializer <- function(name, serializer){
-  if (is.null(.globals$serializers)){
-    .globals$serializers <- list()
-  }
-  if (!is.null(.globals$serializers[[name]])){
-    stop ("Already have a serializer by the name of ", name)
+addSerializer <- function(name, serializer, verbose = TRUE) {
+  if (!is.null(.globals$serializers[[name]])) {
+    if (isTRUE(verbose)) {
+      warning("Overwriting serializer: ", name)
```
Comment (cpsievert) [inline R/serializer.R:17]: Maybe this should be a `message()` (since there are scenarios that are no real user 'mistake')

File: R/serializer.R [hunk 4]
```diff
@@ -17,30 +8,267 @@ NULL
...
+#' filter/endpoint into an HTTP response that can be returned to the client. See
+#' [here](https://book.rplumber.io/articles/rendering-output.html#serializers-1) for
+#' more details on Plumber serializers and how to customize their behavior.
+#' @describeIn serializers Add a static list of headers to each return value. Will add `Content-Disposition` header if a value is the result of `as_attachment()`.
+#' @param ... extra arguments supplied to respective internal serialization function.
+#' @param headers `list()` of headers to add to the response object
+#' @param serialize_fn Function to serialize the data. The result object will be converted to a character string. Ex: `jsonlite::toJSON`.
+#' @export
+serializer_headers <- function(headers, serialize_fn = identity) {
+  stopifnot(is.function(serialize_fn))
```
Comment (cpsievert) [inline R/serializer.R:83]:
```suggestion
serializer_headers <- function(headers = list(), serialize_fn = identity) {
  stopifnot(is.function(serialize_fn))
  stopifnot(is.list(headers))
```

File: R/serializer.R [hunk 5]
```diff
@@ -17,30 +8,267 @@ NULL
...
+  })
+}
+
+#' @describeIn serializers Text serializer. See [format()] for more details.
+#' @export
+serializer_text <- function(...) {
+  serializer_content_type("text/plain; charset=UTF-8", function(val) {
+    format(val, ...)
+  })
+}
```
Comment (cpsievert) [inline R/serializer.R:220]: Consider renaming to `serializer_format()` and adding  `serializer_print()` and maybe `serializer_text()` as `capture.output(cat(x))`

Comment (cpsievert) [reply R/serializer.R:220]: (for consistency with shiny)

File: R/serializer.R (hunk 2 above)
Comment (blairj09) [reply R/serializer.R:80]: `as_attachment` isn't exported 😁

File: R/serializer.R [hunk 6]
```diff
@@ -17,30 +8,267 @@ NULL
...
+#' Return an attachment response
+#'
+#' This will set the appropriate fields in the `Content-Disposition` header value.
+#' To make sure the attachement is used, be sure your serializer eventually calls `serializer_headers`
+#'
+#' @param value Response value to be saved
+#' @param filename File name to use when saving the attachment.
+#'   If no `filename` is provided, the `value` will be treated as a regular attachment
+#' @return Object with class `"plumber_attachment"`
+#' @examples
```
Comment (schloerke) [inline R/serializer.R:47]: Thank you @blairj09 

```suggestion
#' @export
#' @examples
```

**Pull request merged**